### DISCUSSION TITLE
Delay PUBACK till after message processing using QoS 1

### DISCUSSION THREAD

**Discussion by Lacutas82** [dotnet/MQTTnet#1468] [Q&A] [answered]
I'd like to perform some processing on the received message before acknowledging that it's been received correctly.  Is there a way to delay sending of the PUBACK under QoS 1 until after some processing has been completed?

**Comment by chkr1011** [comment, marked as answer]
You can set the "AutoAcknowledge" to false and call "AcknowledgeAsync" after you finishes your stuff.

**Reply by Lacutas82** [reply]
Thank you for this!

**Reply by Rohit03sahu** [reply]
@chkr1011 

I have been looking for the same thing like how can we control the Acknowledge QOS=1.

Could you please provide us any example for mqttnet c# library ?

**Reply by Lacutas82** [reply]
@Rohit03sahu 
I haven't tested this as have been working on something else, but I'm presuming something like this would delay the acknowledgement until you're ready to actually do it.

```
await mqttClient.ConnectAsync(mqttClientOptions);

mqttClient.ApplicationMessageReceivedAsync += delegate(MqttApplicationMessageReceivedEventArgs eventArgs)
{
    eventArgs.AutoAcknowledge = false;
    
    Console.WriteLine(eventArgs.ApplicationMessage.Topic);
    
    eventArgs.AcknowledgeAsync(CancellationToken.None);

    return Task.CompletedTask;
};
```

**Comment by delphikit** [comment]
Would this also be possible with ApplicationMessageSkippedAsync?
